### PR TITLE
[RLlib; docs] New API stack migration guide.

### PR DESCRIPTION
**Pull request by sven1977**
Step-by-step new API stack migration guide.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simonsays1980 — APPROVED** [review]
LSTM. Some suggestions here and there.

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        enable_connectors=True,
+        remote_worker_envs=False
+        remote_env_batch_wait_ms=0,
+        preprocessor_pref="deepmind",
+        enable_tf1_exec_eagerly=False,
+        sampler_perf_stats_ema_coef=None,
+    )
+
+
+In case you were using the `observation_filter` setting, perform the following translations:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/new-api-stack-migration-guide.rst:206]: Doesn't exist yet :(

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+the :ref:`RLModule documentation <rlmodule-guide>` first, then come back to this location here.
+
+Here are also helpful example scripts on `how to write a custom CNN-containing RLModule <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/custom_cnn_rl_module.py>`__
+and `how to write a custom LSTM-containing RLModule <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/custom_lstm_rl_module.py>`__.
+
+Also, there are different options for translating an existing, custom :py:class:`~ray.rllib.models.modelv2.ModelV2` (old API stack)
+to the new API stack's :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`:
+
+1) You lift your ModelV2 code and drop it into a new, custom RLModule class (see the :ref:`RLModule documentation <rlmodule-guide>` for details).
+1) You use an Algorithm checkpoint or a Policy checkpoint that you have from an old API stack training run and use this with the `new stack RLModule convenience wrapper <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/migrate_modelv2_to_new_api_stack_by_policy_checkpoint.py>`__.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/new-api-stack-migration-guide.rst:338]: This is correct. Sphinx will automatically enumerate these.

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 3]
```diff
@@ -7,58 +7,80 @@
...
-Over the next couple of months, the Ray Team will continue to test, benchmark, bug-fix, and
-further polish these new APIs as well as rollout more and more algorithms that you can run in
-either stack.
-The goal is to reach a state where the new stack can completely replace the old one.
+Over the next few months, the RLlib Team will continue to document, test, benchmark, bug-fix, and
+further polish these new APIs as well as rollout more and more algorithms (with a focus on offline RL)
+that you can run in the new stack.
 
-Keep in mind that due to its alpha nature, when using the new stack, you might run into issues and encounter instabilities.
+Keep in mind that due to its alpha nature, when using the new stack, you might run into
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/rllib-new-api-stack.rst:70]: You are right. This is from an older iteration. Will fix this and make it more bullish.

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 4]
```diff
@@ -7,58 +7,80 @@
...
+     - Yes
+     - Yes
 
 
-Over the next couple of months, the Ray Team will continue to test, benchmark, bug-fix, and
-further polish these new APIs as well as rollout more and more algorithms that you can run in
-either stack.
-The goal is to reach a state where the new stack can completely replace the old one.
+Over the next few months, the RLlib Team will continue to document, test, benchmark, bug-fix, and
+further polish these new APIs as well as rollout more and more algorithms (with a focus on offline RL)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/rllib-new-api-stack.rst:67]: Sure, ok, I thought, we'd be more conservative and make sure this all works also for GPU and multi-GPU? Just don't want to announce too much that's not 98ish% stable.

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+.. testcode::
+
+    from ray.rllib.algorithms.ppo import PPOConfig
+
+    config = (
+        PPOConfig()
+        # Switch both the new API stack flags to True (both False by default).
+        # This enables the use of
+        # a) RLModule (replaces ModelV2) and Learner (replaces Policy)
+        # b) the correct EnvRunner (single-agent vs multi-agent) and ConnectorV2 pipelines.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/new-api-stack-migration-guide.rst:39]: done

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+
+The `num_gpus` and `_fake_gpus` settings have been deprecated. In order to place your RLModule on one or more GPUs on the Learner side,
+do the following:
+
+.. testcode::
+
+    # The following setting is equivalent to the old stack's `config.resources(num_gpus=2)`.
+    config.learners(
+        num_learners=2,
+        num_gpus_per_learner=1,
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/new-api-stack-migration-guide.rst:82]: done

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+
+
+Learning Rate- and Coefficient Schedules
+........................................
+
+If you are using schedules for your learning rate or other coefficients (for example the
+`entropy_coeff` setting in PPO), provide scheduling information directly in the respective setting.
+There is no specific, separate setting anymore for scheduling behavior.
+
+When defining a schedule, provide a list of 2-tuples, where the first item is the global timestep
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/new-api-stack-migration-guide.rst:149]: done. And linked to example script.

**Review by angelinalg — COMMENTED** [review]
first batch of suggestions - so for the quantity.

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+Custom Loss Functions and/or Policies
+-------------------------------------
+
+In case you are using one or more custom loss functions and/or custom (PyTorch) optimizers to train your models, instead of doing these
+customizations inside the old stack's Policy class, you need to move these logic into the new API stack's
+:py:class:`~ray.rllib.core.learner.learner.Learner` class.
+
+:ref:`See here for more details on how to write a custom Learner <learner-guide>`.
+
+Here are also helpful example scripts on `how to write a simple custom loss function <https://github.com/ray-project/ray/blob/master/rllib/examples/learners/custom_loss_fn_simple.py>`__
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:400]:
```suggestion
The following example scripts show how to write:
- `a simple custom loss function <https://github.com/ray-project/ray/blob/master/rllib/examples/learners/custom_loss_fn_simple.py>`__
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+and `how to write a custom Learner with 2 optimizers and different learning rates for each <https://github.com/ray-project/ray/blob/master/rllib/examples/learners/separate_vf_lr_and_optimizer.py>`__.
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:401]:
```suggestion
- `a custom Learner with 2 optimizers and different learning rates for each <https://github.com/ray-project/ray/blob/master/rllib/examples/learners/separate_vf_lr_and_optimizer.py>`__.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+
+Note that the Policy class is no longer supported in the new API stack. This class used to hold a
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:403]:
```suggestion
Note that the new API stack doesn't support the Policy class. In the old stack, this class holds a
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+neural network (now moved into :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`),
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:404]:
```suggestion
neural network, which is the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` in the new API stack,
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+a (old stack) connector (now moved into :py:class:`~ray.rllib.connector.connector_v2.ConnectorV2`),
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:405]:
```suggestion
an old stack connector, which is the :py:class:`~ray.rllib.connector.connector_v2.ConnectorV2` in the new API stack,
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+and one or more optimizers and losses (now moved into :py:class:`~ray.rllib.core.learner.learner.Learner`).
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:406]:
```suggestion
and one or more optimizers and losses, which are the :py:class:`~ray.rllib.core.learner.learner.Learner` class in the new API stack.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+
+The RLModule API is much more flexible than the old stack's Policy API and
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:408]:
```suggestion
The RL Module API is much more flexible than the old stack's Policy API and
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+provides a cleaner separation-of-concerns experience (things related to action
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:409]:
```suggestion
provides a cleaner separation-of-concerns experience. Things related to action
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 16]
```diff
@@ -0,0 +1,431 @@
...
+inference run on the EnvRunners, things related to updating run on the Learner workers)
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:410]:
```suggestion
inference run on the EnvRunners, and things related to updating run on the Learner workers
```

File: doc/source/rllib/new-api-stack-migration-guide.rst (hunk 15 above)
Comment (angelinalg) [reply doc/source/rllib/new-api-stack-migration-guide.rst:409]: Consider using a more descriptive word in place of `things`.

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+as well as superior scalability allowing training in a multi-GPU setup in any Ray cluster
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:411]:
```suggestion
It also provides superior scalability, allowing training in a multi-GPU setup in any Ray cluster
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+and multi-node + multi-GPU training on the `Anyscale <https://anyscale.com>`__ platform.
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:412]:
```suggestion
and multi-node with multi-GPU training on the `Anyscale <https://anyscale.com>`__ platform.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+
+
+Custom (old-stack) Connectors
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:415]:
```suggestion
Custom connectors (old-stack) 
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 20]
```diff
@@ -0,0 +1,431 @@
...
+-----------------------------
+
+If you are using custom (old API stack) connectors, you have to lift your logic into the
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:418]:
```suggestion
If you're using custom connectors from the old API stack, move your logic into the
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 21]
```diff
@@ -0,0 +1,431 @@
...
+new :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` API.
+Translate your agent connectors into env-to-module ConnectorV2 pieces and your
+action connectors into module-to-env ConnectorV2 pieces.
+
+The :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` documentation is work in progress and linked from here shortly.
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:423]:
```suggestion
The :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` documentation is under development.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 22]
```diff
@@ -0,0 +1,431 @@
...
+
+In the meantime, take a look at some examples on how to write ConnectorV2 pieces for the
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:425]:
```suggestion
The following are some examples on how to write ConnectorV2 pieces for the
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 23]
```diff
@@ -0,0 +1,431 @@
...
+different pipelines:
+
+1) `Example on how to perform observation frame-stacking <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/frame_stacking.py>`__.
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:428]:
```suggestion
1) `Observation frame-stacking <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/frame_stacking.py>`__.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 24]
```diff
@@ -0,0 +1,431 @@
...
+1) `Example on how to add the most recent action and reward to the RLModule's input <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/prev_actions_prev_rewards.py>`__.
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:429]:
```suggestion
1) `Add the most recent action and reward to the RL Module's input <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/prev_actions_prev_rewards.py>`__.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 25]
```diff
@@ -0,0 +1,431 @@
...
+1) `Example on how to do mean-std filtering on all observations <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/mean_std_filtering.py>`__.
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:430]:
```suggestion
1) `Mean-std filtering on all observations <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/mean_std_filtering.py>`__.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 26]
```diff
@@ -0,0 +1,431 @@
...
+1) `Example on how to flatten any complex observation space to a 1D space <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/flatten_observations_dict_space.py>`__.
```
Comment (angelinalg) [inline doc/source/rllib/new-api-stack-migration-guide.rst:431]:
```suggestion
1) `Flatten any complex observation space to a 1D space <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/flatten_observations_dict_space.py>`__.
```

**Review by angelinalg — COMMENTED** [review]
Rounding out the feedback for new-api-stack-migration-guide.

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 27]
```diff
@@ -7,58 +7,79 @@
 RLlib's New API Stack
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:7]:
```suggestion
RLlib's new API stack
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 28]
```diff
@@ -7,58 +7,79 @@
...
 =====================
 
+.. hint::
+
+    This section describes in detail what the new API stack is and why you should migrate to it
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:12]:
```suggestion
    This section describes the new API stack and why you should migrate to it
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 29]
```diff
@@ -7,58 +7,79 @@
...
+    (in case you have old API stack custom code).
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:13]:
```suggestion
    if you have old API stack custom code.
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 30]
```diff
@@ -7,58 +7,79 @@
...
+     - Yes
 
 
-Over the next couple of months, the Ray Team will continue to test, benchmark, bug-fix, and
-further polish these new APIs as well as rollout more and more algorithms that you can run in
-either stack.
-The goal is to reach a state where the new stack can completely replace the old one.
+Over the next few months, the RLlib Team continues to document, test, benchmark, bug-fix, and
+further polish these new APIs as well as rollout more algorithms
+that you can run in the new stack (with a focus on offline RL).
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:68]:
```suggestion
that you can run in the new stack, with a focus on offline RL.
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 31]
```diff
@@ -7,58 +7,79 @@
...
 
-Keep in mind that due to its alpha nature, when using the new stack, you might run into issues and encounter instabilities.
-Also, rest assured that you are able to continue using your custom classes and setups
-on the old API stack for the foreseeable future (beyond Ray 3.0).
+Also know that you are able to continue using your custom classes and setups
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:70]:
```suggestion
You can continue using custom classes and setups
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 32]
```diff
@@ -7,58 +7,79 @@
...
+on the old API stack for the foreseeable future (beyond Ray 3.0), however, we do recommend you
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:71]:
```suggestion
on the old API stack for the foreseeable future, beyond Ray 3.0. However, you should
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 33]
```diff
@@ -7,58 +7,79 @@
...
+to migrate to the new stack :ref:`using the migration guide here <rllib-new-api-stack-migration-guide>`
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:72]:
```suggestion
migrate to the new stack with the :ref:`migration guide <rllib-new-api-stack-migration-guide>`
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 34]
```diff
@@ -7,58 +7,79 @@
...
 
 
 What is the New API Stack?
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:75]:
```suggestion
New API stack
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 35]
```diff
@@ -78,22 +99,14 @@ The :py:class:`~ray.rllib.algorithm.algorithm_config.AlgorithmConfig` and :py:cl
...
 ---------------------------------
 
-Eventually, all users of RLlib should switch over to running experiments and developing their custom classes
-against the new API stack.
-
-Right now, it's only available for a few algorithms and setups (see table above), however, if you do use
-PPO (single- or multi-agent) or SAC (single-agent), you should try it.
-
-The following section, lists some compelling reasons to migrate to the new stack.
+We recommend to cut over your code from old- to new API stack as soon as possible, as the
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:102]:
```suggestion
Migrate your code from the old to new API stack as soon as possible.
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 36]
```diff
@@ -78,22 +99,14 @@ The :py:class:`~ray.rllib.algorithm.algorithm_config.AlgorithmConfig` and :py:cl
...
+classes and APIs have reached sufficient stability at this point and we only expect very minor changes
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:103]:
```suggestion
The classes and APIs are sufficiently stability. The Ray team expects very minor changes.
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 37]
```diff
@@ -78,22 +99,14 @@ The :py:class:`~ray.rllib.algorithm.algorithm_config.AlgorithmConfig` and :py:cl
...
+from here on.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:104]:
```suggestion
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 38]
```diff
@@ -7,58 +7,79 @@
 RLlib's New API Stack
 =====================
 
+.. hint::
+
+    This section describes in detail what the new API stack is and why you should migrate to it
+    (in case you have old API stack custom code).
+    :ref:`See here for a comprehensive migration guide <rllib-new-api-stack-migration-guide>`.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:14]:
```suggestion
   See the :ref:`migration guide <rllib-new-api-stack-migration-guide>` for details.
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 39]
```diff
@@ -78,22 +99,14 @@ The :py:class:`~ray.rllib.algorithm.algorithm_config.AlgorithmConfig` and :py:cl
...
-Right now, it's only available for a few algorithms and setups (see table above), however, if you do use
-PPO (single- or multi-agent) or SAC (single-agent), you should try it.
-
-The following section, lists some compelling reasons to migrate to the new stack.
+We recommend to cut over your code from old- to new API stack as soon as possible, as the
+classes and APIs have reached sufficient stability at this point and we only expect very minor changes
+from here on.
 
-Note these indicators against using it at this early stage:
+See here for a comprehensive migration guide allowing you to step-by-step translate your code from
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:106]:
```suggestion
See the :ref:`New API stack migration guide <rllib-new-api-stack-migration-guide>` for a comprehensive migration guide with step-by-step instructions on translating your code from the
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 40]
```diff
@@ -78,22 +99,14 @@ The :py:class:`~ray.rllib.algorithm.algorithm_config.AlgorithmConfig` and :py:cl
...
+old- to new API stack in no time.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:107]:
```suggestion
old to new API stack.
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 41]
```diff
@@ -78,22 +99,14 @@ The :py:class:`~ray.rllib.algorithm.algorithm_config.AlgorithmConfig` and :py:cl
...
 
-1) You're using a custom :py:class:`~ray.rllib.models.modelv2.ModelV2` class and aren't interested right now in moving it into the new :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` API.
-1) You're using a custom :py:class:`~ray.rllib.policy.policy.Policy` class (e.g., with a custom loss function and aren't interested right now in moving it into the new :py:class:`~ray.rllib.core.learner.learner.Learner` API.
-1) You're using custom :py:class:`~ray.rllib.connector.connector.Connector` classes and aren't interested right now in moving them into the new :py:class:`~ray.rllib.connector.connector_v2.ConnectorV2` API.
-
-If any of the above applies to you, don't migrate for now, and continue running with the old API stack. Migrate to the new
-stack whenever you're ready to re-write some small part of your code.
+Also, the following section, lists more compelling reasons to migrate to the new stack.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:109]:
```suggestion
A comparison of the old to new API stack provides additional motivation for migrating to the new stack.
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 42]
```diff
@@ -78,22 +99,14 @@ The :py:class:`~ray.rllib.algorithm.algorithm_config.AlgorithmConfig` and :py:cl
...
 
 
 Comparison to the Old API Stack
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:112]:
```suggestion
Comparison to the old API stack
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 43]
```diff
@@ -137,31 +150,6 @@ This table compares features and design choices between the new and old API stac
...
-            :end-before: __enabling-new-api-stack-ma-ppo-end__
-
-
-    .. tab-item:: Single Agent **SAC**
-
-        .. literalinclude:: doc_code/new_api_stack.py
-            :language: python
-            :start-after: __enabling-new-api-stack-sa-sac-begin__
-            :end-before: __enabling-new-api-stack-sa-sac-end__
+:ref:`See here for our complete and comprehensive migration guide <rllib-new-api-stack-migration-guide>`
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:153]:
```suggestion
See :ref:`New API stack migration guide <rllib-new-api-stack-migration-guide>` for a complete and comprehensive migration guide
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 44]
```diff
@@ -137,31 +150,6 @@ This table compares features and design choices between the new and old API stac
...
+describing in detail which steps you have to take and which changes you have to apply to your
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:154]:
```suggestion
with detailed steps and changes to apply to your
```

File: doc/source/rllib/rllib-new-api-stack.rst [hunk 45]
```diff
@@ -137,31 +150,6 @@ This table compares features and design choices between the new and old API stac
...
+custom RLlib classes in order to migrate from the old- to the new stack.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-new-api-stack.rst:155]:
```suggestion
custom RLlib classes to migrate from the old to the new stack.
```

File: rllib/algorithms/algorithm_config.py [hunk 46]
```diff
@@ -1725,6 +1726,17 @@ def env_runners(
                 synchronously in turn with their update step (e.g. PPO or DQN). Not
```
Comment (angelinalg) [inline rllib/algorithms/algorithm_config.py:1726]:
```suggestion
                synchronously in turn with their update step (e.g., PPO or DQN). Not
```

File: rllib/algorithms/algorithm_config.py [hunk 47]
```diff
@@ -1725,6 +1726,17 @@ def env_runners(
...
                 relevant for any algos that sample asynchronously, such as APPO or
                 IMPALA.
+            max_requests_in_flight_per_env_runner: Max number of inflight requests
+                to each EnvRunner worker. See the FaultTolerantActorManager class for
+                more details.
+                Tuning these values is important when running experiments with
+                large sample batches, where there is the risk that the object store may
+                fill up, causing spilling of objects to disk. This can cause any
+                asynchronous requests to become very slow, making your experiment run
+                slow as well. You can inspect the object store during your experiment
```
Comment (angelinalg) [inline rllib/algorithms/algorithm_config.py:1736]:
```suggestion
                slowly as well. You can inspect the object store during your experiment
```

File: rllib/algorithms/algorithm_config.py [hunk 48]
```diff
@@ -1725,6 +1726,17 @@ def env_runners(
...
+                via a call to ray memory on your headnode, and by using the ray
```
Comment (angelinalg) [inline rllib/algorithms/algorithm_config.py:1737]:
```suggestion
                via a call to Ray memory on your head node, and by using the Ray
```

File: rllib/algorithms/algorithm_config.py [hunk 49]
```diff
@@ -3317,8 +3330,9 @@ def experimental(
                 classes or a dictionary mapping module IDs to such a list of respective
                 scheduler classes. Multiple scheduler classes can be applied in sequence
                 and will be stepped in the same sequence as defined here. Note, most
-                learning rate schedulers need arguments to be configured, i.e. you need
-                to partially initialize the schedulers in the list(s).
+                learning rate schedulers need arguments to be configured, i.e. you might
```
Comment (angelinalg) [inline rllib/algorithms/algorithm_config.py:3333]:
```suggestion
                learning rate schedulers need arguments to be configured, that is, you might
```

File: rllib/algorithms/algorithm_config.py [hunk 50]
```diff
@@ -3317,8 +3330,9 @@ def experimental(
...
+                have to partially initialize the schedulers in the list(s) using
+                `functools.partial`.
             _tf_policy_handles_more_than_one_loss: Experimental flag.
                 If True, TFPolicy will handle more than one loss/optimizer.
```
Comment (angelinalg) [inline rllib/algorithms/algorithm_config.py:3337]:
```suggestion
                If True, TFPolicy handles more than one loss or optimizer.
```

File: rllib/examples/learners/ppo_with_torch_lr_schedulers.py [hunk 51]
```diff
@@ -5,9 +5,9 @@
 optimizer. In this way even more complex learning rate schedules can be assembled.
 
 This example shows:
-    - how to partially initialize multiple learning rate schedulers in PyTorch.
-    - how to chain these schedulers together and pass the chain into RLlib's
-        configuration.
+    - how to configure multiple learning rate schedulers (as a chained pipeline) in
```
Comment (angelinalg) [inline rllib/examples/learners/ppo_with_torch_lr_schedulers.py:8]:
```suggestion
    - how to configure multiple learning rate schedulers, as a chained pipeline, in
```

**Review by angelinalg — APPROVED** [review]
Sorry for the delay.

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 52]
```diff
@@ -0,0 +1,430 @@
...
+
+Learning rate- and coefficient schedules
+........................................
+
+If you're using schedules for learning rate or other coefficients, for example, the
+`entropy_coeff` setting in PPO, provide scheduling information directly in the respective setting.
+Scheduling behavior doesn't require a specific, separate setting anymore.
+
+When defining a schedule, provide a list of 2-tuples, where the first item is the global timestep
+(*num_env_steps_sampled_lifetime* in the reported metrics) and the second item is the value that should
```
Comment (sven1977) [inline doc/source/rllib/new-api-stack-migration-guide.rst:172]:
```suggestion
(*num_env_steps_sampled_lifetime* in the reported metrics) and the second item is the value that the learning rate should reach at that timestep.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 53]
```diff
@@ -0,0 +1,430 @@
...
+be reached at that timestep. Always start the first 2-tuple with timestep 0. Note that RLlib linearly interpolates values between
```
Comment (sven1977) [inline doc/source/rllib/new-api-stack-migration-guide.rst:173]:
```suggestion
Always start the first 2-tuple with timestep 0. Note that RLlib linearly interpolates values between
```

**Pull request merged**